### PR TITLE
Migrate Tooltip component + Fix styles

### PR DESCRIPTION
**Pull request by stuniel**

### CODE REVIEW

File: packages/tooltip/README.md [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+- `react-dom: ^16.2.0`
+
+These packages are required by `@talixo/tooltip`, but you have to install them manually,
+to avoid having different versions of these in your application.
+
+## Supported props
+
+Property name | Type                               | Default | Description                    
+--------------|------------------------------------|:-------:|-----------------------
+children      |	node	                             | n/a     | Tooltipped elements
```
Comment (rangoo94) [inline packages/tooltip/README.md]: Tab characters breaks it :)

File: packages/tooltip/src/Tooltip.js [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+
+    return (
+      <div
+        className={`${name}-hover`}
+        onClick={this.handleMouseClick}
+        onMouseEnter={this.handleMouseEnter}
+        onMouseLeave={this.handleMouseLeave}
+        onMouseOver={this.handleMouseOver}
+      >
+        <div ref={node => (this.el = node)} style={{ display: 'inline-block' }} className={`${name}-wrapper`}>
```
Comment (rangoo94) [inline packages/tooltip/src/Tooltip.js]: Could you extract this `ref` into method?

File: packages/tooltip/src/Tooltip.js [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+ * @param {string} [props.children]
+ * @param {string} [props.className]
+ * @param {string} [props.color]
+ * @param {string} [props.fade]
+ * @param {string} [props.fadeTime]
+ * @param {string} [props.isOpen]
+ * @param {string} [props.style]
+ * @param {string} [props.position]
+ * @param {string} [props.render]
+ * @returns {React.Element}
```
Comment (rangoo94) [inline packages/tooltip/src/Tooltip.js]: When you use `class` these params are not correct. It should be passed to `constructor`; `@returns` could be passed to `render`

File: packages/tooltip/src/Tooltip.js [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+  componentWillReceiveProps (nextProps) {
+    if (this.props.isOpen !== nextProps.isOpen) this.setState({ isOpen: nextProps.isOpen })
+  }
+
+  updatePosition () {
+    const { top, left } = getPositionNearElement(this.el, this.props.position)
+    this.setState({ top, left })
+  }
+
+  handleMouseEnter () {
```
Comment (rangoo94) [inline packages/tooltip/src/Tooltip.js]: Do you need both `mouseenter and `mouseover` events?

File: packages/tooltip/src/Tooltip.js [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+    if (!this.state.clicked && !this.state.isOpen) this.setState({ isOpen: true })
+  }
+
+  handleMouseClick () {
+    if (!_.isUndefined(this.props.isOpen)) return
+    this.setState({ clicked: true, isOpen: false })
+  }
+
+  render () {
+    const {
```
Comment (rangoo94) [inline packages/tooltip/src/Tooltip.js]: It could be smaller :)

File: packages/tooltip/src/Tooltip.js (hunk 3 above)
Comment (rangoo94) [inline packages/tooltip/src/Tooltip.js]: When you use `class` these params are not correct. It should be passed to `constructor`; `@returns` could be passed to `render`

File: packages/tooltip/src/Tooltip.js [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+      rootNode,
+      style
+    } = this.props
+
+    const defaultFadeTime = 600
+    const transition = fade ? { transition: `opacity ${fadeTime || defaultFadeTime}ms` } : null
+
+    return (
+      <div
+        className={`${name}-hover`}
```
Comment (rangoo94) [inline packages/tooltip/src/Tooltip.js]: `--hover` may be better (for BEM)

File: packages/tooltip/src/Tooltip.js [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+  fade: PropTypes.bool,
+
+  /** Fade time */
+  fadeTime: PropTypes.number,
+
+  /** Controls whether tooltip is open */
+  isOpen: PropTypes.bool,
+
+  /** Tooltip position */
+  position: PropTypes.oneOf(['left', 'right', 'top', 'bottom']).isRequired,
```
Comment (rangoo94) [inline packages/tooltip/src/Tooltip.js]: Couldn't be `top` as default? It's pretty common

File: packages/tooltip/src/Tooltip.js (hunk 4 above)
Comment (rangoo94) [inline packages/tooltip/src/Tooltip.js]: Do you need both `mouseenter` and `mouseover` events?

File: packages/tooltip/src/Tooltip.js [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+
+  /** Renders toolyip content */
```
Comment (rangoo94) [inline packages/tooltip/src/Tooltip.js]: typo

File: packages/tooltip/src/Tooltip.js [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+import { CSSTransition, TransitionGroup } from 'react-transition-group'
+
+import { getPositionNearElement } from '../utils/position'
+import { prefix } from '@talixo/commons'
+
+const name = prefix('tooltip')
+
+class Portal extends React.Component {
+  constructor (props) {
+    super(props)
```
Comment (rangoo94) [inline packages/tooltip/src/Tooltip.js]: It's not handling root change

File: packages/tooltip/src/Tooltip.js [hunk 10]
```diff
@@ -0,0 +1,194 @@
...
+  fadeTime: PropTypes.number,
+
+  /** Controls whether tooltip is open */
+  isOpen: PropTypes.bool,
+
+  /** Tooltip position */
+  position: PropTypes.oneOf(['left', 'right', 'top', 'bottom']).isRequired,
+
+  /** Renders toolyip content */
+  render: PropTypes.func,
```
Comment (rangoo94) [inline packages/tooltip/src/Tooltip.js]: Right now `render` is required

File: packages/tooltip/src/Tooltip.js (hunk 8 above)
Comment (rangoo94) [inline packages/tooltip/src/Tooltip.js]: typo

File: packages/tooltip/utils/position.js [hunk 11]
```diff
@@ -0,0 +1,31 @@
+export const getPositionNearElement = (element, position) => {
+  const boundingClientRect = element.getBoundingClientRect()
```
Comment (rangoo94) [inline packages/tooltip/utils/position.js]: What if position of element changes? Right now no events are checked

File: packages/tooltip/src/Tooltip.js [hunk 12]
```diff
@@ -0,0 +1,215 @@
...
+import { getPositionNearElement } from '../utils/position'
+import { prefix } from '@talixo/commons'
+
+const name = prefix('tooltip')
+
+class Portal extends React.Component {
+  constructor (props) {
+    super(props)
+    this.state = {
+      root: this.props.root || document.querySelector('body')
```
Comment (rangoo94) [inline packages/tooltip/src/Tooltip.js]: Do you really need to keep it in state?

File: packages/tooltip/src/Tooltip.js [hunk 13]
```diff
@@ -0,0 +1,215 @@
...
+
+  handleMouseClick () {
+    if (!_.isUndefined(this.props.isOpen)) return
+    this.setState({ clicked: true, isOpen: false })
+  }
+
+  setRef (node, child) {
+    this.el = node
+    const {ref} = child
+    if (typeof ref === 'function') {
```
Comment (rangoo94) [inline packages/tooltip/src/Tooltip.js]: Why do you need that?

File: packages/tooltip/src/Tooltip.js [hunk 14]
```diff
@@ -0,0 +1,215 @@
...
+    return (
+      <div
+        className={`${name}-hover`}
+        onClick={this.handleMouseClick}
+        onMouseEnter={this.handleMouseEnter}
+        onMouseLeave={this.handleMouseLeave}
+        onMouseOver={this.handleMouseOver}
+      >
+        {React.Children.map(this.props.children, child =>
+          React.cloneElement(React.Children.only(child), {
```
Comment (rangoo94) [inline packages/tooltip/src/Tooltip.js]: When you are using `React.Children.map` isn't `child` always the only child there?

File: packages/tooltip/src/Tooltip.js [hunk 15]
```diff
@@ -0,0 +1,215 @@
...
+
+    return (
+      <div
+        className={`${name}-hover`}
+        onClick={this.handleMouseClick}
+        onMouseEnter={this.handleMouseEnter}
+        onMouseLeave={this.handleMouseLeave}
+        onMouseOver={this.handleMouseOver}
+      >
+        {React.Children.map(this.props.children, child =>
```
Comment (rangoo94) [inline packages/tooltip/src/Tooltip.js]: Could you extract some of these above for clearer code?

File: packages/tooltip/stories.js [hunk 16]
```diff
@@ -0,0 +1,105 @@
...
+        backgroundColor: '#eee',
+        padding: '1rem',
+        cursor: 'pointer'
+      }}
+    >
+      Tooltip right
+    </span>
+  </Tooltip>
+))
+addStory('controlled', readme, () => <ControlledTooltip />)
```
Comment (rangoo94) [inline packages/tooltip/stories.js]: could you use `addStory.controlled` to have problem source code visible in Storybook?

File: packages/tooltip/styles/geometry.sass [hunk 17]
```diff
@@ -0,0 +1,92 @@
+@import './config'
+
+@include module('tooltip')
+  position: absolute
+  display: inline-block
+  padding: grid-unit(1) grid-unit(1.5)
+  border: $border-width solid
```
Comment (rangoo94) [inline packages/tooltip/styles/geometry.sass]: `border: $border-width solid $color-orange-talixo` and you can get rid of `border-color` below

File: packages/tooltip/styles/geometry.sass [hunk 18]
```diff
@@ -0,0 +1,92 @@
...
+    opacity: 1
+
+  &-fade-exit-active
+    opacity: 0.01
+
+  &--right
+    transform: translateY(-50%) translateX(grid-unit(1))
+
+    &:before
+      transform: translateY(-50%) translateX(50%) rotate(-45deg) skew(20deg, 20deg)
```
Comment (rangoo94) [inline packages/tooltip/styles/geometry.sass]: What is this `skew` for?

Comment (rangoo94) [reply packages/tooltip/styles/geometry.sass]: As I understand you are creating triangles - could you use it with just standard border triangles?
http://apps.eky.hk/css-triangle-generator/

File: packages/tooltip/src/Tooltip.js [hunk 19]
```diff
@@ -0,0 +1,213 @@
...
+                    [`${name}--${color}`]: color !== undefined,
+                    [`${name}--${position}`]: position !== undefined
+                  })}
+                  style={{
+                    top: this.state.top,
+                    left: this.state.left,
+                    ...transition,
+                    ...style
+                  }}
+                >
```
Comment (maciek15) [inline packages/tooltip/src/Tooltip.js]: `className` and `style` could be extracted above

File: packages/tooltip/README.md [hunk 20]
```diff
@@ -0,0 +1,39 @@
...
+- `react-dom: ^16.2.0`
+
+These packages are required by `@talixo/tooltip`, but you have to install them manually,
+to avoid having different versions of these in your application.
+
+## Supported props
+
+Property name | Type                               | Default | Description
+--------------|------------------------------------|:-------:|-----------------------
+children      |	node	                             | n/a     | Tooltipped elements
```
Comment (michalochman) [inline packages/tooltip/README.md]: Tab characters

**Pull request merged**